modP2pImpl/test/org/aion/p2p/impl1/tasks/TaskConnectPeersTest.java

```diff
@@ -0,0 +1,289 @@
+                        e.printStackTrace();
+                    }
+                    itor.remove();
+                }
+            }
+        }
+    }
+
+    @Before
+    public void setup() throws IOException {
+        MockitoAnnotations.initMocks(this);
+        Map<String, String> logMap = new HashMap<>();
+        logMap.put(LogEnum.P2P.name(), LogLevel.TRACE.name());
+        AionLoggerFactory.init(logMap);
+
+        System.setProperty("java.net.preferIPv4Stack", "true");
+        ssc = ServerSocketChannel.open();
+        assertNotNull(ssc);
+        ssc.configureBlocking(false);
+        ssc.socket().setReuseAddress(true);
+        port = 50000 + r.nextInt(15535);
+        ssc.socket().bind(new InetSocketAddress(port));
+        // Create the selector
+        selector = Selector.open();
+        assertNotNull(selector);
+        ssc.register(selector, SelectionKey.OP_ACCEPT);
+
+        listen = new ThreadTCPServer(selector);
+        assertNotNull(listen);
+        listen.start();
+    }
+
+    @After
+    public void tearDown() throws IOException, InterruptedException {
+        listen.interrupt();
+        Thread.sleep(1000);
+        ssc.close();
+    }
+
+
+    @Test
+    public void testRun() throws InterruptedException {
+        AtomicBoolean atb = new AtomicBoolean(true);
+        TaskConnectPeers tcp = new TaskConnectPeers(p2pMgr, atb, nodeMgr, 128, selector, sendMsgQue,
+            rhs);
+        assertNotNull(tcp);
+
+        Thread t = new Thread(tcp);
+        t.start();
+        assertTrue(t.isAlive());
+        Thread.sleep(10);
+        atb.set(false);
+        Thread.sleep(2000);
+        assertEquals("TERMINATED", t.getState().toString());
+    }
+
+    @Test
+    public void testRun1() throws InterruptedException {
+        AtomicBoolean atb = new AtomicBoolean(true);
+        TaskConnectPeers tcp = new TaskConnectPeers(p2pMgr, atb, nodeMgr, 128, selector, sendMsgQue,
+            rhs);
+        assertNotNull(tcp);
+
+        when(nodeMgr.activeNodesSize()).thenReturn(128);
+        when(nodeMgr.tempNodesTake()).thenReturn(null);
+
+        when(node.getIdHash()).thenReturn(1);
+        when(node.getPort()).thenReturn(port);
+        when(node.getIpStr()).thenReturn("127.0.0.1");
+        when(nodeMgr.notAtOutboundList(node.getIdHash())).thenReturn(true);
+        when(nodeMgr.notActiveNode(node.getIdHash())).thenReturn(true);
+
+        Thread t = new Thread(tcp);
+        t.start();
+        assertTrue(t.isAlive());
+        Thread.sleep(10);
+
+        // should see the loop continue every sec
+        Thread.sleep(2000);
+
+        when(nodeMgr.activeNodesSize()).thenReturn(127);
+        // should see the loop continue every sec due to null node been taken
+        Thread.sleep(2000);
+
+        when(node.getIdShort()).thenReturn("1");
+        when(nodeMgr.tempNodesTake()).thenReturn(node);
+        when(node.getIfFromBootList()).thenReturn(true);
+
+        Thread.sleep(2000);
```

Comment:
Why do you need sleeps in between setting up mocks and setting the value of an AtomicBoolean?

Reply:
it is not a good test, but the idea is to change the mock condition to verify&flow the thread logic